src/log.c

```diff
@@ -562,11 +658,14 @@ qd_error_t qd_log_entity(qd_entity_t *entity)
         //
         QD_ERROR_BREAK();
 
-        //
-        // Obtain all attributes from the entity before obtaining the log_source_lock.
-        // We do this because functions like qd_entity_get_string and qd_entity_get_bool ultimately call qd_vlog_impl() which
-        // also holds the log_source_lock when calling write_log().
-        //
+        qd_log_source_t *src = qd_log_source(module); /* The original(already existing) log source */
+
+        // 1956: Since we are messing with this source's sink, here we
+        //       can use the local sink-lock.
+        //       Previously, we needed to obtain these attributes outside of the lock, because
+        //       of a deadlock issue.
+
+        sys_mutex_lock(src->sink_lock);
```

Comment:
I don't think this is correct: we shouldn't be protecting the sink because the sink is not changing.  The source is being modified - and if any sink attributes need to be modified we simply create a new one and drop the reference to the old sink (which can still be in use by other sources).

Reply:
For this I think maybe a per-source read/write lock would be a better choice:  take the read lock while logging.  Take the write lock during this call.